react/client/src/contexts/RoutingContext.tsx

```diff
@@ -11,18 +13,61 @@ const RoutingContextProvider = ({
   children,
   history,
 }: RoutingProviderProps) => {
-  const url = history.location.pathname;
-  const pageNumber = Number(url.slice(url.indexOf('/form') + 6)) || 0;
+  const [formSteps, setFormSteps] = useState([FORM_STEPS.NONE]);
+  const [currentStepIdx, setCurrentStepIdx] = useState(0);
+  const currentStep = formSteps[currentStepIdx];
+  const { pathname } = history.location;
+
+  const navigateToFormPage = (newPageUrl: string) => {
+    history.push(`/form/${newPageUrl}`);
+  };
 
   const goNextPage = () => {
-    history.push(`/form/${pageNumber + 1}`);
+    const nextStep = getNextFormStep(currentStep);
+    setCurrentStepIdx(currentStepIdx + 1);
+    setFormSteps([...formSteps, nextStep]);
+    const nextPageUrl = PAGES[nextStep];
+    navigateToFormPage(nextPageUrl);
   };
+
   const goBackPage = () => {
-    history.push(`/form/${pageNumber - 1}`);
+    const prevStepIdx = Math.max(currentStepIdx - 1, 0);
+    const prevStep = formSteps[prevStepIdx];
+    setCurrentStepIdx(prevStepIdx);
+    setFormSteps(formSteps.slice(0, prevStepIdx + 1));
+    const prevPageUrl = PAGES[prevStep];
+    navigateToFormPage(prevPageUrl);
   };
 
+  // handle arriving directly from a url
+  useEffect(() => {
+    const pathMatcher = pathname.match(/(?<=\/form\/)\w*/) || [];
+    const stepFromPathName = pathMatcher[0] || '';
+
+    // redirect back to the first page when accessing another random page
+    // (in the future we would first check what data is currently cached before
+    // deciding if we redirect or not)
+    if (stepFromPathName !== PAGES[FORM_STEPS.NONE] && formSteps.length <= 1) {
+      // setCurrentStepIdx(0);
+      // setFormSteps([FORM_STEPS.NONE]);
+      // navigateToFormPage(PAGES[FORM_STEPS.NONE]);
+    }
```

Comment:
The one caveat with my overengineered implementation is that direct changes to the url in the browser needs to be handled. Both directly accessing a url and pressing the back button on the browser. The url issue is handled here but I'll have to think about the back button.

Reply:
Yes, though I can't imagine there will be many cases where the user will be manipulating the url while using the app - but at least we have a fallback to the home page for now.